### PR TITLE
[WebAssembly] Support assembly parsing for new EH

### PR DESCRIPTION
**Pull request by aheejin**
This adds assembly parsing support for the new EH (exnref) proposal.

`try_table` parsing is a little tricky because catch clause lists use `()` and the multivalue block return types also use `()`. This handles all combinations below:
- No return type (void) + no catch list
- No return type (void) + catch list
- Single return type + no catch list
- Single return type + catch list
- Multivalue return type + no catch list
- Multivalue return type + catch list

This does not include AsmTypeCheck support yet. That's the reason why this adds a new test file and use `--no-type-check` in the command line. After the type checker is added as a follow-up, I plan to merge https://github.com/llvm/llvm-project/blob/main/llvm/test/MC/WebAssembly/eh-assembly-legacy.s with this file. (Turning on `-mattr=+exception-handling` adds support for all legacy and new EH instructions in the assembly. `-wasm-enable-exnref` in `llc` only controls which instructions to generate and it doesn't affect `llvm-mc` and assembly parsing.)

### CODE REVIEW

**Review by dschuff — COMMENTED** [review]
Any thoughts on what might be the most likely error cases to arise in practice? I noticed there are no tests of error cases; I don't think I'd necessarily want to try to be really comprehensive on covering every possible error cases but if we think some categories are more likely, it might be good to have something.

File: llvm/lib/Target/WebAssembly/AsmParser/WebAssemblyAsmParser.cpp [hunk 1]
```diff
@@ -69,12 +69,23 @@ struct WebAssemblyOperand : public MCParsedAsmOperand {
     std::vector<unsigned> List;
   };
 
+  struct CaLOpElem {
+    int64_t Opcode;
+    const MCExpr *Tag;
+    int64_t Dest;
```
[the remark this answers is not in the document]
Comment (aheejin) [reply llvm/lib/Target/WebAssembly/AsmParser/WebAssemblyAsmParser.cpp]: I made them `int64_t` just because the token's parser's return value was `int64_t`:
https://github.com/llvm/llvm-project/blob/e99755d41c8d5ee8594fa161ac05fbef02d902aa/llvm/include/llvm/MC/MCAsmMacro.h#L115-L118

And yeah I agree giving them the real type is better. Changed `Opcode` to `uint8_t` and `Dest` to `unsigned` to reflect the real binary spec.

**Comment by aheejin** [comment]
> Any thoughts on what might be the most likely error cases to arise in practice? I noticed there are no tests of error cases; I don't think I'd necessarily want to try to be really comprehensive on covering every possible error cases but if we think some categories are more likely, it might be good to have something.

Added some error cases.

Tried to check an invalid tag too, but that's a realm of AsmTypeCheck.

File: llvm/lib/Target/WebAssembly/MCTargetDesc/WebAssemblyInstPrinter.cpp [hunk 2]
```diff
@@ -378,7 +378,7 @@ void WebAssemblyInstPrinter::printCatchList(const MCInst *MI, unsigned OpNo,
     const MCSymbolRefExpr *TagExpr = nullptr;
     const MCSymbolWasm *TagSym = nullptr;
     assert(Op.isExpr());
-    TagExpr = dyn_cast<MCSymbolRefExpr>(Op.getExpr());
+    TagExpr = cast<MCSymbolRefExpr>(Op.getExpr());
```
Comment (aheejin) [inline llvm/lib/Target/WebAssembly/MCTargetDesc/WebAssemblyInstPrinter.cpp:381]: drive-by fix

**Review by dschuff — APPROVED** [review]
